In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.formula.api as smf

# PS 07
David Gao
## 1 Heart attack: inferential modeling
#### 1.1 

In [46]:
heart = pd.read_csv(r"heart.csv", sep=",")
heart.head(5)

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [3]:
heart.shape

(303, 14)

#### 1.2

In [4]:
heart.isna().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

In [5]:
heart.dtypes

age           int64
sex           int64
cp            int64
trtbps        int64
chol          int64
fbs           int64
restecg       int64
thalachh      int64
exng          int64
oldpeak     float64
slp           int64
caa           int64
thall         int64
output        int64
dtype: object

In [6]:
heart.describe()

age         sex          cp      trtbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg    thalachh        exng     oldpeak         slp         caa  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

            thall      output  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [7]:
heart[heart.output == 1].output.sum()/303 # calculating percentage of heart attack

0.5445544554455446

Categorical variables are sex, cp, fbs, restecg, exng, and output. And 54.45 percentage of these patients have heart attack. The differences between data and ducumentation is that cp is categorized as 0,1, 2, 3 instead of 1, 2, 3, 4. And caa is 0-4 instead of 0-3.
#### 1.3
Nominal variables are sex, cp, fbs, restecg, esng, caa and output. And caa might be ordinal variables.
#### 1.4

In [10]:
m1 = smf.logit("output ~ age + C(sex) + C(cp) + trtbps + chol + C(fbs) + C(restecg) + thalachh + C(exng) + caa + slp + oldpeak + thall", data = heart).fit()
m1.summary()

Optimization terminated successfully.
         Current function value: 0.345478
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 output   No. Observations:                  303
Model:                          Logit   Df Residuals:                      286
Method:                           MLE   Df Model:                           16
Date:                Sat, 28 May 2022   Pseudo R-squ.:                  0.4987
Time:                        21:30:37   Log-Likelihood:                -104.68
converged:                       True   LL-Null:                       -208.82
Covariance Type:            nonrobust   LLR p-value:                 1.674e-35
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           3.2475      2.577      1.260      0.208      -1.804       8.299
C(sex)[T.1]        -1.7661      0.477     -3.699      0.000      -2.702      -0.830
C(cp)[T.1]          1.1925      0.544      2.192      0.028       0.126       2.259
C(cp)[T.2]          2.0373      0.464      4.386      0.000       1.127       2.948
C(cp)[T.3]          2.1519      0.633      3.400      0.001       0.911       3.392
C(fbs)[T.1]        -0.0111      0.538     -0.021      0.984      -1.066       1.044
C(restecg)[T.1]     0.4970      0.363      1.370      0.171      -0.214       1.208
C(restecg)[T.2]    -0.3395      2.227     -0.152      0.879      -4.704       4.025
C(exng)[T.1]       -0.9026      0.416     -2.169      0.030      -1.718      -0.087
age                -0.0038      0.023     -0.163      0.871      -0.049       0.042
trtbps             -0.0178      0.010     -1.707      0.088      -0.038       0.003
chol               -0.0049      0.004     -1.274      0.203      -0.012       0.003
thalachh            0.0220      0.011      2.097      0.036       0.001       0.043
caa                -0.7898      0.194     -4.069      0.000      -1.170      -0.409
slp                 0.5783      0.355      1.630      0.103      -0.117       1.274
oldpeak            -0.4956      0.220     -2.250      0.024      -0.927      -0.064
thall              -0.8875      0.292     -3.040      0.002      -1.460      -0.315
===================================================================================
"""

#### 1.5

In [11]:
m1.get_margeff().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
        Logit Marginal Effects       
=====================================
Dep. Variable:                 output
Method:                          dydx
At:                           overall
===================================================================================
                     dy/dx    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
C(sex)[T.1]        -0.1920      0.048     -4.013      0.000      -0.286      -0.098
C(cp)[T.1]          0.1297      0.057      2.266      0.023       0.018       0.242
C(cp)[T.2]          0.2215      0.044      4.983      0.000       0.134       0.309
C(cp)[T.3]          0.2340      0.064      3.674      0.000       0.109       0.359
C(fbs)[T.1]        -0.0012      0.059     -0.021      0.984      -0.116       0.114
C(restecg)[T.1]     0.0540      0.039      1.383      0.167      -0.023       0.131
C(restecg)[T.2]    -0.0369      0.242     -0.152      0.879      -0.511       0.438
C(exng)[T.1]       -0.0981      0.044     -2.237      0.025      -0.184      -0.012
age                -0.0004      0.003     -0.163      0.871      -0.005       0.005
trtbps             -0.0019      0.001     -1.736      0.083      -0.004       0.000
chol               -0.0005      0.000     -1.285      0.199      -0.001       0.000
thalachh            0.0024      0.001      2.150      0.032       0.000       0.005
caa                -0.0859      0.019     -4.533      0.000      -0.123      -0.049
slp                 0.0629      0.038      1.652      0.098      -0.012       0.137
oldpeak            -0.0539      0.023     -2.322      0.020      -0.099      -0.008
thall              -0.0965      0.030     -3.226      0.001      -0.155      -0.038
===================================================================================
"""

#### 1.6
#### (a)
People that is one year older have 0.04 percentage low in probability to get a heart attack. This variable is not statistically significant as 0.871 > 0.05.
#### (b)
People with sex equals 1, that are males, have 19.2 percentage lower in probability to get heart attack. And this variable is statistically significant. Sex equals 0, female, is the reference category.
#### (c)
For cp variable, the cp = 0 category is used as reference category, though unclear which kind of chest pain this category is corresponding to as there is no such category in variable specification. For the logistic regression all cp categories are statistically significant. And as the pain category goes from 0 to 3, the chance for having heart attack also increases.
#### (d)
Sex = 1, fbs = 1, restecg = 2, exng = 1, age, trtbps, caa, oldpeak, and thall are variables associated with lower chance of heart attack. And thall, oldpeak, caa, exng, and sex = 1 are statistically significant. These kind of do not make sense, as these variable seems to be unrelated to heart attack.
#### 1.7

In [12]:
heart["age_category"] = pd.cut(heart.age,
                                      bins = [ 0, 40, 55, 65, 100], 
                                      labels = ["0 - 40", "40 - 55", "55 - 65", "65 - 100"],
                                      right = False)
heart.head(3)

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   

   caa  thall  output age_category  
0    0      1       1      55 - 65  
1    0      2       1       0 - 40  
2    0      2       1      40 - 55

In [13]:
m2 = smf.logit("output ~ C(age_category)", data = heart).fit()
m2.get_margeff().summary()

Optimization terminated successfully.
         Current function value: 0.643033
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
        Logit Marginal Effects       
=====================================
Dep. Variable:                 output
Method:                          dydx
At:                           overall
===============================================================================================
                                 dy/dx    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
C(age_category)[T.40 - 55]     -0.0700      0.137     -0.511      0.609      -0.338       0.198
C(age_category)[T.55 - 65]     -0.3650      0.131     -2.786      0.005      -0.622      -0.108
C(age_category)[T.65 - 100]    -0.2368      0.146     -1.624      0.104      -0.523       0.049
===============================================================================================
"""

Age group 0 to 40 is the reference category.
#### 1.8
The heart attack probability is falling in age. And the reason why is that this data only includes who had heart attack before and those got killed by heart attack directly is not inclueded in this data. And old people who had heart attack is more likely got killed thus not included in this dataset, which means the old poeple in this data are more likely to be those who had no heart attack 
#### 1.9
#### (a)

In [14]:
prediction_prob = m1.predict()
prediction_prob[:10]

array([0.71802931, 0.71976376, 0.96909372, 0.92888982, 0.80630678,
       0.7465016 , 0.86069741, 0.86779617, 0.82969012, 0.93854501])

#### (b)

In [15]:
prediction = prediction_prob < 0.5
prediction[:10]

array([False, False, False, False, False, False, False, False, False,
       False])

#### (c)

In [16]:
newData = pd.DataFrame({"age": 60, "sex": 1, "cp": 2, "trtbps": 125, "chol": 250, "fbs": 0, "restecg": 2, "thalachh": heart.thalachh.mean(), "exng": 0,
                        "oldpeak": heart.oldpeak.mean(), "slp": heart.slp.mean(), "caa": heart.caa.mean(), "thall": heart.thall.mean()},
                       index = [0])
newData

age  sex  cp  trtbps  chol  fbs  restecg    thalachh  exng   oldpeak  \
0   60    1   2     125   250    0        2  149.646865     0  1.039604   

       slp       caa     thall  
0  1.39934  0.729373  2.313531

In [17]:
m1.predict(newData)

0    0.613006
dtype: float64

AS 0.613 > 0.5, this 60 year old man is likely to have heart attack.

#### (d)

In [104]:
cm = pd.crosstab(heart.output, prediction+0)
cm

col_0     0    1
output          
0        30  108
1       151   14

In [122]:
from sklearn.metrics import accuracy_score
accuracy_score(heart.output, prediction)

0.14521452145214522

The accuracy is approximately 0.125.

## Heart atttack: predictive modeling
#### 2.1

In [50]:
y = heart.output
X = heart.drop(["output"], axis = 1)

# create dummies for categorical variables
X["sex"] = pd.cut(X.sex, 
                  bins = [ 0, 1, np.inf], 
                  labels = ["0", "1"],
                  right = False)
X["cp"] = pd.cut(X.cp, 
                  bins = [ 0, 1, 2, 3, np.inf], 
                  labels = ["0", "1", "2", "3"],
                  right = False)
X["fbs"] = pd.cut(X.fbs, 
                  bins = [ 0, 1, np.inf], 
                  labels = ["0", "1"],
                  right = False)
X["restecg"] = pd.cut(X.restecg, 
                  bins = [ 0, 1, 2, np.inf], 
                  labels = ["0", "1", "2"],
                  right = False)
X["exng"] = pd.cut(X.exng, 
                  bins = [ 0, 1, np.inf], 
                  labels = ["0", "1"],
                  right = False)
X = pd.get_dummies(X, drop_first=True)
X.head(5)

age  trtbps  chol  thalachh  oldpeak  slp  caa  thall  sex_1  cp_1  cp_2  \
0   63     145   233       150      2.3    0    0      1      1     0     0   
1   37     130   250       187      3.5    0    0      2      1     0     1   
2   41     130   204       172      1.4    2    0      2      0     1     0   
3   56     120   236       178      0.8    2    0      2      1     1     0   
4   57     120   354       163      0.6    2    0      2      0     0     0   

   cp_3  fbs_1  restecg_1  restecg_2  exng_1  
0     1      1          0          0       0  
1     0      0          1          0       0  
2     0      0          0          0       0  
3     0      0          1          0       0  
4     0      0          1          0       1

Reference: [drop function](https://www.geeksforgeeks.org/how-to-drop-one-or-multiple-columns-in-pandas-dataframe/)

In [51]:
from sklearn.linear_model import LogisticRegression
m = LogisticRegression(max_iter=5000, penalty="none")
_ = m.fit(X, y)

In [52]:
m.intercept_

array([3.19637128])

In [53]:
m.coef_

array([[-0.00350836, -0.01773841, -0.004883  ,  0.02209152, -0.49545646,
         0.58591037, -0.79465733, -0.88537004, -1.76269472,  1.16598096,
         2.02789266,  2.12706734,  0.01653439,  0.50594008, -0.28838175,
        -0.91577647]])

#### 2.2

In [58]:
prob = m.predict_proba(X)
prob[:10, 1]

array([0.71813945, 0.71854052, 0.96843302, 0.92858398, 0.80753809,
       0.74850025, 0.85790339, 0.86714357, 0.83557893, 0.93927212])

#### 2.3

In [92]:
outcome = m.predict(X) == 1
outcome[:10]

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

All first 10 people have heart attack.
#### 2.4

In [93]:
outcome_prob = prob[:, 1] > 0.5
outcome_prob = outcome_prob
outcome_prob[:10]

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

In [102]:
((outcome + 0) - (outcome_prob + 0)).sum()

0

The results are the same in both methods
#### 2.5

In [103]:
cm = pd.crosstab(heart.output, outcome + 0)
cm

col_0     0    1
output          
0       108   30
1        14  151

#### 2.6

In [109]:
precision = 151/(151+30)
print("The precision is", precision, ".")

The precision is 0.8342541436464088 .


In [110]:
recall = 151/(151 + 14)
print("The recall is", recall, ".")

The recall is 0.9151515151515152 .


In [126]:
from sklearn.metrics import accuracy_score
print("Accuracy is", accuracy_score(heart.output, outcome+0))

Accuracy is 0.8547854785478548


In [128]:
from sklearn.metrics import f1_score
print("F-score is", f1_score(heart.output, outcome+0))

F-score is 0.8728323699421965


I think we should imporve accuracy, as we do not want both false positive and false negative situations. We want to give treatments to thoes who really have heart attack accrately, with out giving people with no heart attack any treatment.
#### 2.7
Assume simple model predict everyone have heart attack. Then TN value is 0, and the FP value is the number of poeple with no heart attack. And as all people are predicted as having heart attack, then FN is also 0. And TP is the number of true number of poeple that have heart attack.

In [112]:
(heart.output == 0).sum() # number of FP, people with no heart attack

138

In [113]:
(heart.output == 1).sum() # number of TP, people with heart attack

165

In [116]:
cm = pd.DataFrame({"predict N": [0, 0], "predict P": [138, 165]}, index = ["true N", "true P"])
cm # confusion matrix

predict N  predict P
true N          0        138
true P          0        165

TN, FP\
FN, TP

In [117]:
# Accuracy
print("Accuracy is", 165/303)

Accuracy is 0.5445544554455446


In [118]:
# Precision
print("Precision is", 165/(138 + 165))

Precision is 0.5445544554455446


In [119]:
# Recall
print("Recall is", 165/(0 + 165))

Recall is 1.0


I spent approximately 6 hours on this PS.